File: examples/lh-server/typescript/src/04-conditionals/main.ts
import { createTaskWorker, LHConfig, Workflow } from "littlehorse-client";
import { z } from "zod";
import { closeOnShutdown } from "../config.js";

function fetchContactMethod(userId: string): string {
  return ["obiwan", "padme", "satine"].includes(userId)
    ? "COMLINK"
    : "HOLOGRAM";
}

function sendComlink(userId: string, message: string): string {
  const result = `sent comlink ${message} to user ${userId}`;
  console.log(result);
  return result;
}

function sendHologram(userId: string, message: string): string {
  const result = `sent hologram ${message} to user ${userId}`;
  console.log(result);
  return result;
}

const config = LHConfig.from({});
const workers = [
  createTaskWorker(fetchContactMethod, "fetch-contact-method", config, {
    inputVars: { userId: z.string() },
    outputSchema: z.string(),
  }),
  createTaskWorker(sendComlink, "send-comlink-message", config, {
    inputVars: { userId: z.string(), message: z.string() },
    outputSchema: z.string(),
  }),
  createTaskWorker(sendHologram, "send-hologram", config, {
    inputVars: { userId: z.string(), message: z.string() },
    outputSchema: z.string(),
  }),
];

const workflow = Workflow.newWorkflow("send-message", (wf) => {
  const userId = wf.declareStr("user-id").required();
  const message = wf.declareStr("message").required();
  const preferredContact = wf.declareStr("contact-method");

  preferredContact.assign(wf.execute("fetch-contact-method", userId));
  wf.doIfElse(
    preferredContact.isEqualTo("COMLINK"),
    (ifHandler) => {
      ifHandler.execute("send-comlink-message", userId, message);
    },
    (elseHandler) => {
      elseHandler.execute("send-hologram", userId, message);
    },
  );
});

await closeOnShutdown(workers);
await Promise.all(workers.map((worker) => worker.registerTaskDef()));
await workflow.registerWfSpec(config);
console.log('Run with: lhctl run send-message user-id anakin message "Hello there"');
await Promise.all(workers.map((worker) => worker.start()));

File: examples/lh-server/python/04-conditionals/main.py
import asyncio
import sys

import littlehorse
from littlehorse.config import LHConfig
from littlehorse.worker import LHTaskWorker
from tasks import fetch_contact_method, send_comlink_message, send_hologram
from workflow import get_workflow


def register(config: LHConfig) -> None:
    littlehorse.create_task_def(fetch_contact_method, "fetch-contact-method", config)
    littlehorse.create_task_def(send_comlink_message, "send-comlink-message", config)
    littlehorse.create_task_def(send_hologram, "send-hologram", config)
    littlehorse.create_workflow_spec(get_workflow(), config)


async def workers(config: LHConfig) -> None:
    await littlehorse.start(
        LHTaskWorker(fetch_contact_method, "fetch-contact-method", config),
        LHTaskWorker(send_comlink_message, "send-comlink-message", config),
        LHTaskWorker(send_hologram, "send-hologram", config),
    )


if __name__ == "__main__":
    config = LHConfig()
    if len(sys.argv) != 2 or sys.argv[1] not in {"register", "workers"}:
        raise SystemExit("Usage: main.py [register|workers]")
    register(config) if sys.argv[1] == "register" else asyncio.run(workers(config))
